File: common-java/src/main/java/se/dandel/lednotification/Event.java
package se.dandel.lednotification;

import org.apache.commons.lang.builder.ToStringBuilder;
import org.apache.commons.lang.builder.ToStringStyle;

public class Event {

    private EventType type;

    private EventSource source;

    private EventPriority priority;

    public Event(EventSource source, EventType type) {
        this.source = source;
        this.type = type;
    }

    public Event(EventSource source, EventType type, EventPriority priority) {
        this(source, type);
        this.priority = priority;
    }

    public EventSource getSource() {
        return source;
    }

    public EventType getType() {
        return type;
    }

    public void setSource(EventSource source) {
        this.source = source;
    }

    public void setType(EventType type) {
        this.type = type;
    }

    public EventPriority getPriority() {
        return priority;
    }

    public void setPriority(EventPriority priority) {
        this.priority = priority;
    }

    public static Event newAlert(EventSource source) {
        return new Event(source, EventType.NEW_ALERT);
    }

    public static Event cancelAlert(EventSource source) {
        return new Event(source, EventType.CANCEL_ALERT);
    }

    public static Event newNotification(EventSource source, EventPriority priority) {
        return new Event(source, EventType.NEW_NOTIFICATION, priority);
    }

    public static Event cancelNotification(EventSource source, EventPriority priority) {
        return new Event(source, EventType.CANCEL_NOTIFICATION, priority);
    }

    public static Event ping(EventSource source) {
        return new Event(source, EventType.PING);
    }

    public static Event pong(EventSource source) {
        return new Event(source, EventType.PONG);
    }

    @Override
    public String toString() {
        return ToStringBuilder.reflectionToString(this, ToStringStyle.SHORT_PREFIX_STYLE);
    }

}

File: cplusplus/src/Event.cpp
#include <map>
#include "spark_wiring_usbserial.h"
#include "Event.h"

namespace lednotification {

    map< std::string, EventType> eventTypeDictionary = {
        {"NEW_ALERT", EventType::NEW_ALERT},
        {"CANCEL_ALERT", EventType::CANCEL_ALERT},
        {"NEW_NOTIFICATION", EventType::NEW_NOTIFICATION},
        {"CANCEL_NOTIFICATION", EventType::CANCEL_NOTIFICATION},
        {"PING", EventType::PING},
        {"PONG", EventType::PONG}
    };

    map< std::string, EventPriority> eventPriorityDictionary = {
        {"CRITICAL", EventPriority::CRITICAL},
        {"HIGH", EventPriority::HIGH},
        {"MEDIUM", EventPriority::MEDIUM},
        {"LOW", EventPriority::LOW}
    };

    EventType lookupEventType(std::string s) {
      Serial.println("Looking up event type");
      Serial.println(s.c_str());
      map< std::string, EventType >::iterator i = eventTypeDictionary.find(s);
      if( i != eventTypeDictionary.end()) {
          return i->second;
      }
    }

    EventPriority lookupEventPriority(std::string s) {
      Serial.println("Looking up event priority");
      Serial.println(s.c_str());
      map< std::string, EventPriority >::iterator i = eventPriorityDictionary.find(s);
      if( i != eventPriorityDictionary.end()) {
          return i->second;
      }
    }

    Event::Event(std::string s, std::string t) {
      Serial.println("Howdy");
      source = s;
      type = lookupEventType(t);
    }

    Event::Event(std::string s, std::string t, std::string p) {
      source = s;
      type = lookupEventType(t);
      priority = lookupEventPriority(p);
    }
}

